### ISSUE TITLE
Watcher causes stuck build for `paraglide-js compile`

### ISSUE THREAD

**Issue by NilsJacobsen** [inlang/monorepo#1590]
## Problem 
See the problem in my loom ->
https://www.loom.com/share/b424a6b0d1f44da08b299911409a23d1?sid=d5511d0f-fe7d-4d7f-a9d2-072d4fff5b55

## Insight
We have consumers of the loadProjectApi that want to have a watched file system and there are consumers that doesn't. (e.g. **watcher-friends:** editor, ide-extension; **no watcher needed:** badge, cli, paraglide-js compile)

## Proposal
We want to have the watcher logic in the SDK to keep it super easy for the inlang Apps. But for some use cases the Apps want an inlang project that does not react on changes. 

We could add a not required prop to the loadProject API. 

Cons: 
- Extending the easy API feel bad

Pro: 
- We don't want Apps to integrate the logic of watcher

```ts
export const loadProject = async (args: {
	settingsFilePath: string
	nodeishFs: NodeishFilesystemSubset
	_import?: ImportFunction
	_capture?: (id: string, props: Record<string, unknown>) => void
	_watch?: boolean  //default true
}): Promise<InlangProject> => {
        ...
}
```

If you have an easier proposal, please share. I would love not to extend the API. 
@inlang/sdk @inlang/paraglide-js

**Comment by felixhaeberle** [comment]
Ultimately, we need to control if the process should "watch & keep alive" or "build & complete". There is no other possibility than implementing a `_watch` or `_watchFiles`to differentiate between those functionalities.

Unfortunately, you can't just simply run a watcher in the background. This would mean running a hidden process which then also needs lifecycle control (hard) in the control of an inlang app.

**Comment by samuelstroschein** [comment]
@inlang/lix @NilsJacobsen does the `nodeisFs.watch` API implicitly create a long-running process? 

I expect the compile command to run through without getting stuck if `fs.watch` is called. After all, the program, not the SDK, should determine if the process is long running or not.

**Comment by NilsJacobsen** [comment]
Here's the update:
https://www.loom.com/share/71f5290ad4934ace96726b1c4ea8eef1?sid=b534320a-da48-40d0-86ed-e2254b0a4ab6

- Wrapper works
- Tests are passing

**Problem:**
the watch API wants me to stop the process manually and I do not know when I should do that.
@inlang/lix @samuelstroschein @felixhaeberle

**Comment by NilsJacobsen** [comment]
Update:

@janfjohannes adds a persistent option to watcher. That should solve it. Now I'm gonna try to test it in editor and ide extension.

**Comment by felixhaeberle** [comment]
@NilsJacobsen wuhu 🎉 could you test it with #1518 ?

**Comment by samuelstroschein** [comment]
@NilsJacobsen @janfjohannes is the "persistent option" in the watcher node fs API compliant? 

I don't quite seem to understand why the process is long running. I would have expected that `fs.watch()` is a subscription that automatically disposes if the process is exited, but not blocking the process from exiting

**Comment by NilsJacobsen** [comment]
> is the "persistent option" in the watcher node fs API compliant?

![image](https://github.com/inlang/monorepo/assets/58360188/01546bcc-2248-4a96-a8ee-044af8c6732a)


Update: 

I tested it with the editor. In general, it works. We used to manually `setMessages`. This is not necessary anymore.

**Problem:** Plugins load and save the whole ast. When the SDK gets notified we can not only read from the changed path. We need to go through the plugin. That means watching for 60 + files leads in x-times loadMessages.

**Comment by samuelstroschein** [comment]
> Problem: Plugins load and save the whole ast. When the SDK gets notified we can not only read from the changed path. We need to go through the plugin. That means watching for 60 + files leads in x-times loadMessages.

Will be fixed with #1585 and a "directory proposal" I bring forward at the end of the week/early next week. For now, I want to complete paraglide 1.0

**Comment by NilsJacobsen** [comment]
So it is working in the editor (for small and medium size), ide-extension does not fail, but has a type error (@felixhaeberle is looking into this), we can merge that tomorrow morning and work on the vite plugin.👍
